app/src/test/java/com/android/streetworkapp/model/workout/WorkoutRepositoryFirestoreTest.kt

```diff
@@ -0,0 +1,182 @@
+    `when`(documentRef.get()).thenReturn(Tasks.forResult(document))
+
+    repository.addOrUpdateWorkoutSession(uid, workoutSession)
+
+    verify(documentRef).update(eq("workoutSessions"), eq(listOf(workoutSession)))
+  }
+
+  @Test
+  fun deleteWorkoutSessionRemovesSession() = runTest {
+    val uid = "testUid"
+    val sessionToDelete = WorkoutSession("sessionId", 123L, 0L, SessionType.ALONE)
+    val remainingSession = WorkoutSession("remainingSessionId", 456L, 0L, SessionType.ALONE)
+
+    val existingWorkoutData = WorkoutData(uid, listOf(sessionToDelete, remainingSession))
+
+    // Create a list of session maps that mimics the structure retrieved from Firestore
+    val sessionMaps =
+        existingWorkoutData.workoutSessions.map { session ->
+          mapOf(
+              "sessionId" to session.sessionId,
+              "startTime" to session.startTime,
+              "endTime" to session.endTime,
+              "sessionType" to session.sessionType.name,
+              "participants" to session.participants,
+              "exercises" to
+                  session.exercises.map { exercise ->
+                    mapOf(
+                        "name" to exercise.name,
+                        "reps" to exercise.reps,
+                        "sets" to exercise.sets,
+                        "weight" to exercise.weight,
+                        "duration" to exercise.duration)
+                  },
+              "winner" to session.winner)
+        }
+
+    // Mock the Firestore interactions
+    `when`(db.collection("workoutData")).thenReturn(collection)
+    `when`(collection.document(uid)).thenReturn(documentRef)
+    `when`(documentRef.get()).thenReturn(Tasks.forResult(document))
+    `when`(document.exists()).thenReturn(true)
+    `when`(document.id).thenReturn(uid)
+    `when`(document.get("workoutSessions")).thenReturn(sessionMaps)
+
+    // Perform the deletion
+    repository.deleteWorkoutSession(uid, sessionToDelete.sessionId)
+
+    // Capture the updated sessions list
+    val captor = argumentCaptor<List<WorkoutSession>>()
+    verify(documentRef).update(eq("workoutSessions"), captor.capture())
+
+    println("Captured argument: ${captor.firstValue}")
+
+    // Verify that only the remaining session is present
+    assertEquals(listOf(remainingSession), captor.firstValue)
+  }
+
+  @Test
+  fun updateWorkoutSessionDetailsUpdatesDetails() = runTest {
+    val uid = "testUid"
+    val sessionId = "sessionId"
+    val exercises = listOf(Exercise(name = "Push-up"))
+    val endTime = 123456789L
+
+    repository.updateWorkoutSessionDetails(uid, sessionId, exercises, endTime)
+
+    verify(documentRef)
+        .update(
+            mapOf(
+                "workoutSessions.$sessionId.exercises" to exercises,
+                "workoutSessions.$sessionId.endTime" to endTime))
+  }
+
+  @Test
+  fun updateExerciseUpdatesCorrectFieldPath() = runTest {
+    val uid = "testUid"
```

Comment:
There are many val and when that appear multiple time in every test, I think some of them can be put in `setup`. Like  val uid = "testUid" for example